src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,151 @@
+        return elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndexOutOfBoundException(index);
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexOutOfBoundException(index);
+        T removedElement = elements[index];
+        fastRemoveElement(index);
+        return removedElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int indexRemovedElement = getElementIndex(element);
+        T removedElement = elements[getElementIndex(element)];
+        fastRemoveElement(indexRemovedElement);
+        return removedElement;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    @Override
+    public Iterator<T> iterator() {
+        return new Iterator<T>() {
+            private int index = 0;
+            @Override
+            public boolean hasNext() {
+                return (index < elements.length && elements[index] != null);
+            }
+
+            @Override
+            public T next() {
+                if (!hasNext()) {
+                    throw new NoSuchElementException();
+                }
+                return elements[index++];
+            }
+        };
+    }
+
+    private void growIfFuel() {
+        if (elements.length == 1 && elements.length == size) {
+            T[] newArray = (T[])new Object[DEFAULT_CAPACITY];
+            System.arraycopy(elements, 0, newArray, 0, size);
+            elements = newArray;
+        }
+        if (elements.length == size) {
+            int newCapacity = (int)(elements.length * GROWTH_FACTOR);
+            T[] newArray = (T[]) new Object[newCapacity];
+            System.arraycopy(elements, 0, newArray, 0, size);
+            elements = newArray;
+        }
+    }
+
+    private void checkIndexOutOfBoundException(int index) {
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "Index: " + index + " out of array size: " + size);
+        }
+    }
+
+    private int getElementIndex(T element) {
+        int index = -1;
+        for (int i = 0; i < size; i++) {
+            if (elements[i] != null
+                    && elements[i].equals(element)
+                    || elements[i] == element) {
+                index = i;
+                break;
+            }
+        }
+        if (index == -1) {
+            throw new NoSuchElementException("Can't find element: " + element);
+        }
+        return index;
+    }
+
+    private void fastRemoveElement(int index) {
+        if (index == elements.length - 1) {
+            elements[size - 1] = null;
+        } else {
+            System.arraycopy(elements,index + 1, elements, index, size - index - 1);
```

Comment:
How do you think, what happens if you delete 3 time in the row from center position? What will your array contains? Do you need to clear lans position after each copy, or not?

Reply:
You are right, unused objects will remain in memory. I missed this moment


> How do you think, what happens if you delete 3 time in the row from center position? What will your array contains? Do you need to clear lans position after each copy, or not?